services/data_sources/json_notification_data_source.rb

```diff
@@ -0,0 +1,162 @@
+      else
+        if !params[:user] && !params[:provider_credential]
+          raise "Either user or a provider credential is mandatory."
+        else
+          params[:provider_credential] ||= params[:user].provider_credential(type: ProviderCredential::PROVIDER_CREDENTIAL_TYPES[:github])
+          @git_repo = FastlaneCI::GitRepo.new(git_config: self.json_folder_path, provider_credential: params[:provider_credential])
+        end
+      end
+
+      reload_notifications
+    end
+
+    # Returns an array of notifications from the notifications JSON file stored
+    # in the configuration git repo
+    #
+    # @return [Array[Notification]]
+    def notifications
+      JSONProjectDataSource.projects_file_semaphore.synchronize do
+        path = git_repo.file_path("notifications.json")
+        return [] unless File.exist?(path)
+
+        return JSON.parse(File.read(path)).map(&Notification.method(:from_json!))
+      end
+    end
+
+    # Writes the notifications array to the git repo as JSON and commit them as
+    # the CI user
+    #
+    # @param  [Array[Notification]] notifications
+    def notifications=(notifications)
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        File.write(notifications_file_path, JSON.pretty_generate(notifications.map(&:to_object_dictionary)))
+        git_repo.commit_changes!
+      end
+    end
+
+    # Returns `true` if the notification exists in the in-memory notifications object
+    #
+    # @param  [String] id
+    # @return [Boolean]
+    def notification_exist?(id: nil)
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        existing_notification = @notifications.select { |notification| notification.id == id }.first
+        return existing_notification.nil? ? false : true
+      end
+    end
+
+    # Swaps the old notification record with the updated notification record if
+    # the notification exists
+    #
+    # @param  [Notification] notification
+    def update_notification!(notification: nil)
+      notification.updated_at = Time.now
+
+      notification_index = nil
+      existing_notification = nil
+
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        @notifications.each.with_index do |old_notification, index|
+          if old_notification.id == notification.id
+            notification_index = index
+            existing_notification = old_notification
+            break
+          end
+        end
+      end
+
+      if existing_notification.nil?
+        error_message = "Couldn't update notification #{notification.name} because it doesn't exist"
+        logger.debug(error_message)
+        raise error_message
+      else
+        @notifications[notification_index] = notification
+        self.notifications = @notifications
+        logger.debug("Updating notification #{existing_notification.name}, writing out notifications.json to #{notifications_file_path}")
+      end
+    end
```

Comment:
I think this method is a little hard to read. I have a _personal_ aversion to conditionals and temporary variables :) so this is more of a suggestion. I think it can be distilled into something like this:

```ruby
# guard against a non-operation
unless @notifications.any?{|n| n.id == notification.id}
  raise "Couldn't update notification #{notification.name} because it doesn't exist"
end

JSONNotificationDataSource.file_semaphore.synchronize do
  @notifications.map! do |old|
    if old.id == notification.id
      notification
    else
      old
    end
  end
  # todo: persist @notifications here
end
```

Reply:
Again, I was trying to keep consistency with the [`json_project_data_source`](https://github.com/fastlane/ci/blob/master/services/config_data_sources/json_project_data_source.rb#L176-L184) implementation.

If you'd like, I could change it in both locations.

Reply:
@AndrewMcBurney Thanks for pointing that out. A very valid reason! My vote would be to keep consistency.